lib/puppet/parser/compiler.rb

```diff
@@ -854,7 +854,8 @@ def set_node_parameters
 
   def create_settings_scope
     settings_type = Puppet::Resource::Type.new :hostclass, "settings"
-    environment.known_resource_types.add(settings_type)
+    # use replace instead of add to avoid duplication check and illegal merge (PUP-5954)
+    environment.known_resource_types.replace_settings(settings_type)
```

Comment:
I don't think we want this commit d8d2c9be as it causes an error instead of a deprecation warning. See https://tickets.puppetlabs.com/browse/PUP-5954